### PR TITLE
chore: Validate table names locally first

### PR DESCRIPTION
**Pull request by msfstef**
Fixes https://github.com/electric-sql/electric/issues/1814

We still go to PG to validate the table as we need its OID, so this is not strictly necessary, but since we have validation logic we might as well validate things locally to avoid going to PG for clearly invalid table identifiers?

This also fixes the accepted characters by PG for identifiers [based on the docs](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS) as it was excluding some unicode chars and dollar signs before.

### CODE REVIEW

**Comment by balegas** [comment]
That is a cool little thing. Nice!

**Review by robacourt — APPROVED** [review]

File: packages/sync-service/lib/electric/postgres/identifiers.ex [hunk 1]
```diff
@@ -65,7 +98,52 @@ defmodule Electric.Postgres.Identifiers do
...
+      [table] ->
+        case parse(table) do
+          {:ok, parsed} -> {:ok, {"public", parsed}}
+          {:error, reason} -> {:error, reason}
+        end
+
+      [_, _] = idents ->
+        Enum.reduce_while(idents, {:ok, {}}, fn ident, {:ok, schema} ->
+          case parse(ident) do
+            {:ok, parsed} -> {:cont, {:ok, schema |> Tuple.append(parsed)}}
```
Comment (kevin-dp) [inline packages/sync-service/lib/electric/postgres/identifiers.ex:139]: Do we need `Tuple.append` or can we write `{schema, parsed}` ?

File: packages/sync-service/lib/electric/postgres/identifiers.ex [hunk 2]
```diff
@@ -2,6 +2,30 @@ defmodule Electric.Postgres.Identifiers do
   @namedatalen 63
   @ascii_downcase ?a - ?A
 
+  defmodule StringSplitter do
+    @moduledoc """
+    Utility module for splitting strings on a schema delimiter
+    """
+    def split_outside_quotes(string) do
```
Comment (kevin-dp) [inline packages/sync-service/lib/electric/postgres/identifiers.ex:9]: Can we put a typespec on this public function

File: packages/sync-service/lib/electric/postgres/identifiers.ex [hunk 3]
```diff
@@ -65,7 +98,52 @@ defmodule Electric.Postgres.Identifiers do
...
+  def parse_relation(ident) do
+    case StringSplitter.split_outside_quotes(ident) do
+      [table] ->
+        case parse(table) do
+          {:ok, parsed} -> {:ok, {"public", parsed}}
+          {:error, reason} -> {:error, reason}
+        end
+
+      [_, _] = idents ->
+        Enum.reduce_while(idents, {:ok, {}}, fn ident, {:ok, schema} ->
```
Comment (kevin-dp) [inline packages/sync-service/lib/electric/postgres/identifiers.ex:137]: Do we need a `reduce_while` if there are only 2 identifiers?
Could we use a with statement instead?

```ex
[schema, tbl] = idents ->
  with {:ok, schema} <- parse(schema)
       {:ok, table} <- parse(tbl)
  do
    {:ok, {schema, table}}
  else
    err -> err
```

**Review by kevin-dp — APPROVED** [review]
Nice work!
I left a few minor comments.

File: packages/sync-service/lib/electric/postgres/identifiers.ex (hunk 3 above)
Comment (msfstef) [reply packages/sync-service/lib/electric/postgres/identifiers.ex:137]: Yeah I think this is a clearer approach - will do this

**Pull request merged**